contrib/tokens-sanitizer/README.md

```diff
@@ -0,0 +1,103 @@
+# Tokens Sanitizer
+
+This script is designed to sanitize a documents by replacing special tokens with properly spaced versions.
+
+Why would you want to do that?
+Usually, when tokenizing data for pretraining, we split any special token that occur naturally in the text before adding special tokens to denote end of documents.
+This is useful for cases when you want a model to be able to answer questions such as "what is <|endoftext|>?";
+if you don't escape special tokens properly, the model would conflate different uses of `<|endoftext|>`, and likely would never learn its meta-semantics (since it is vastly more likely to be used to indicate the end of a document).
+
+However, in some cases, data pipeline might inject special tokens that should be treated as such:
+For example, when concatenating files in a code repository, you might want to use `<|repo_name|>author/repo<|file_sep|>` to prefix all your code documents.
+
+In that case, we use the following strategy:
+
+1. We first run this sanitization script to process all original documents; if used as described below, this script would take care of modifying special tokens in a way that is equivalent to them being escaped.
+2. Then, we perform whatever text substitution we want to inject special tokens.
+3. Finally, we tokenize, but, this time, we set `split_special_tokens=False`, since sanitization script already took care of encoding.
+
+**⚠️ IMPORTANT ⚠️**
+
+You must use a **tokenizer that has been modified** to support sanitization. In the rest of this Readme, we use [`allenai/dolma2-tokenizer-U10F0F0`](https://huggingface.co/allenai/dolma2-tokenizer-U10F0F0).
+
+
+## How does the sanitization work?
+
+When running the sanitization script, we inject a Unicode code point from [Supplementary Private Use Area-B (SPUA-B)](https://en.wikipedia.org/wiki/Private_Use_Areas).
+The Unicode consortium does not assign symbols to these code points; rather, they are left open for private entities to use in their own software.
+While the Basic Private Use Area (U+E000 to U+F8FF) is commonly used (for example, Apple uses `U+F8FF` for the Apple logo ), the supplementary area B is exceedingly rare.
+I picked `U+10F0F0` for sanitization.
+
+In order for sanitization to work, we must use a tokenizer that is designed to split of this token as part of its pre-tokenization strategy.
+In HuggingFace Tokenizers library, we achieve that by adding the follow pre-tokenization rule:
```

Comment:
Question: there has been talk of moving to a different tokenizer in the future. Will we be able to preserve this behavior in the options we're considering?

Reply:
yes! we would need to extend future tokenizers with this pretokenization rule, though.